pkg/blobserver/diskpacked/diskpacked.go

```diff
@@ -761,3 +771,77 @@ func (m blobMeta) String() string {
 func (m blobMeta) SizedRef(br blob.Ref) blob.SizedRef {
 	return blob.SizedRef{Ref: br, Size: m.size}
 }
+
+type fileRefCount struct {
+	*os.File
+	refCount int
+}
+type fileHandleCache struct {
+	m       map[string]*fileRefCount
+	mu      sync.RWMutex
+	maxSize int
+}
+
+func newFileHandleCache(maxSize int) *fileHandleCache {
+	if maxSize == 0 {
+		maxSize = 512
+	}
+	return &fileHandleCache{m: make(map[string]*fileRefCount, maxSize)}
+}
+
+func (fhc *fileHandleCache) Get(fn string) (*fileRefCount, error) {
+	fhc.mu.RLock()
+	fh := fhc.m[fn]
+	fhc.mu.RUnlock()
+	if fh != nil {
+		fh.refCount++
+		return fh, nil
+	}
+	fhc.mu.Lock()
+	defer fhc.mu.Unlock()
+	fd, err := os.Open(fn)
+	if err != nil {
+		return nil, err
+	}
+	fh = &fileRefCount{File: fd, refCount: 1}
+	fhc.m[fn] = fh
+	return fh, nil
+}
+func (fhc *fileHandleCache) Put(fh *fileRefCount) {
```

Comment:
I may just be unfamiliar with the naming for similar datastructures, but what does "put" mean here?

I see that it decreases the refcount and possibly remove the file handle from the cache.

Looking at the code it seems that the maximum number of file handles isn't really enforced?

What if they all have a refcount of 1?  What is going to stop us from caching an unlimited of file handles if the user does not read a single blob?